Framework/src/QualitiesToTRFCollectionConverter.cxx

```diff
@@ -0,0 +1,134 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file    QualitiesToTRFCollectionConverter.cxx
+/// \author  Piotr Konopka
+///
+
+#include "QualityControl/QualitiesToTRFCollectionConverter.h"
+
+namespace o2::quality_control::core
+{
+
+const char* noQualityObjectsComment = "No Quality Objects found within the specified time range";
+
+QualitiesToTRFCollectionConverter::QualitiesToTRFCollectionConverter(std::string trfcName, std::string detectorCode, uint64_t startTimeLimit, uint64_t endTimeLimit, std::string qoPath)
+  : mStartTimeLimit(startTimeLimit),
+    mEndTimeLimit(endTimeLimit),
+    mQOPath(qoPath),
+    mConverted(new TimeRangeFlagCollection(trfcName, detectorCode)),
+    mCurrentStartTime(0),
+    mCurrentEndTime(startTimeLimit),
+    mQOsIncluded(0),
+    mWorseThanGoodQOs(0)
+{
+}
+
+void QualitiesToTRFCollectionConverter::operator()(const QualityObject& newQO)
+{
+  if (mConverted->getDetector() != newQO.getDetectorName()) {
+    throw std::runtime_error("The TRFCollection '" + mConverted->getName() +
+                             "' expects QOs from detector '" + mConverted->getDetector() +
+                             "' but received a QO for '" + newQO.getDetectorName() + "'");
+  }
+
+  mQOsIncluded++;
+  if (newQO.getQuality().isWorseThan(Quality::Good)) {
+    mWorseThanGoodQOs++;
+  }
+
+  uint64_t validFrom = strtoull(newQO.getMetadata("Valid-From").c_str(), nullptr, 10);
+  uint64_t validUntil = strtoull(newQO.getMetadata("Valid-Until").c_str(), nullptr, 10);
+  if (validFrom < mCurrentStartTime) {
+    throw std::runtime_error("The currently provided QO is dated as earlier than the one before (" //
+                             + std::to_string(validFrom) + " vs. " + std::to_string(mCurrentStartTime) +
+                             "). QOs should be provided to the QualitiesToTRFCollectionConverter in the chronological order");
+  }
+
+  // Is the beginning of time range covered by the first QO provided?
+  if (mCurrentStartTime < mStartTimeLimit && validFrom > mStartTimeLimit) {
+    mConverted->insert({ mStartTimeLimit, validFrom - 1, FlagReasonFactory::MissingQualityObject(), noQualityObjectsComment, mQOPath });
+  }
+
+  mCurrentStartTime = validFrom < mStartTimeLimit ? mStartTimeLimit : validFrom;
+  mCurrentEndTime = validUntil > mEndTimeLimit ? mEndTimeLimit : validUntil;
+
+  if (!mCurrentTRF.has_value() && newQO.getQuality().isWorseThan(Quality::Good)) { //todo check if no QO in some period before the first one arrives
```

Comment:
I don't really undesrtand the todo.

Reply:
this is my bad habit of forgetting to remove "todo"s after they are done.